ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -1,324 +1,763 @@
+-- then spawn a thread that uses @r@, and finally leave the scope of
+-- 'withRegistry', thereby deallocating @r@ -- leaving the thread to run with
+-- a now deallocated resource.
+--
+-- It is /only/ safe for threads to use a given registry if the lifetime of
+-- those threads is tied to the lifetime of the registry. There would be no
+-- problem with the example above if the thread would be terminated when we
+-- exit the scope of 'withRegistry'.
+--
+-- The 'forkThread' combinator provided by the registry therefore does two
+-- things: it allocates the thread as a resource in the registry, so that it can
+-- kill the thread when releasing all resources in the registry. It also records
+-- the thread ID in a set of known threads. Whenever the registry is accessed
+-- from a thread /not/ in this set, the registry throws a runtime exception. The
+-- intention is that this guards against dangerous patterns like the one above.
+--
+-- = Linking
+--
+-- When thread A spawns thread B using 'withAsync', the lifetime of B is tied
+-- to the lifetime of A:
+--
+-- > withAsync $ \threadB -> ..
+--
+-- After all, when A exits the scope of the 'withAsync', thread B will be
+-- killed. The reverse is however not true: thread B can terminate before
+-- thread A. It is often useful for thread A to be able to declare a dependency
+-- on thread B: if B somehow fails, that is, terminates with an exception, we
+-- want that exception to be rethrown in thread A as well. A can achieve this
+-- by /linking/ to B:
+--
+-- > withAsync $ \threadB -> do
+-- >   link threadB
+-- >   ..
+--
+-- Linking is often useful to create a hierarchy of threads that should all
+-- be terminated if one of them fails. It is however less useful if B's lifetime
+-- is not tied to A's. For example, if A does
+--
+-- > threadB <- async $ ..
+-- > link threadB
+--
+-- and A terminates before B does, any exception thrown by B might be send to
+-- a thread that no longer exists. This is particularly problematic when we
+-- start chaining threads: if A spawns-and-links-to B which spawns-and-links-to
+-- C, and C throws an exception, both A and B should be notified, but this will
+-- not be the case if the B link is missing because B has already terminated.
+--
+-- For this reason, the registry's 'linkToRegistry' combinator does not link the
+-- specified to the thread calling 'linkToRegistry', but rather to the thread
+-- that created the registry. After all, the lifetime of threads spawned with
+-- 'forkThread' can certainly exceed the lifetime of their parent threads, but
+-- the lifetime of /all/ threads spawned using the registry will be limited by
+-- the scope of that registry, and hence the lifetime of the thread that created
+-- it. So, when we call 'linkToRegistry', the exception will be thrown the
+-- thread that created the registry, which (if not caught) will cause that that
+-- to exit the scope of 'withRegistry', thereby terminating the entire hierarchy
```

Comment:
Addressing the first point: renamed `newRegistry` to `unsafeNewRegistry`. Also added a section "Combinators primarily for testing" in the exports, and move `unsafeNewRegistry`, `closeRegistry` as well as `countResources` to that section.

Reply:
As regards terminology, I'd rather we assume less rather than more, and be explicit where needed. I've added a note:

```
-- Note on terminology: when thread A forks thread B, we will say that thread A
-- is the " parent " and thread B is the " child ". No further relationship
-- between the two threads is implied by this terminology. In particular, note
-- that the child may outlive the parent. We will use "fork" and "spawn"
-- interchangeably.
```

I agree that "hierarchy" is indeed misleading, and I've rephrased the text to avoid it (details below).

Reply:
I've weakened the "chaining" paragraph to say:

```
-- Linking a parent to a child is however of limited value if the lifetime of
-- the child is not limited by the lifetime of the parent. For example, if A
-- does
--
-- > threadB <- async $ ..
-- > link threadB
--
-- and A terminates before B does, any exception thrown by B might be send to a
-- thread that no longer exists. This is particularly problematic when we start
-- chaining threads: if A spawns-and-links-to B which spawns-and-links-to C, and
-- C throws an exception, perhaps the intention is that this gets rethrown to B,
-- and then rethrown to A, terminating all three threads; however, if B has
-- terminated before the exception is thrown, C will throw the exception to a
-- non-existent thread and A is never notified.
```

Reply:
Avoiding another use of hierarchy:

```
-- thread that created the registry, which (if not caught) will cause that that
-- to exit the scope of 'withRegistry', thereby terminating all threads in that
-- registry.
```

Reply:
The last use of "hierachy" is justified:

```
-- This /also/ includes nested calls to 'withRegistry'. Since the lifetime of
-- such a registry (and all resources within) is tied to the thread calling
-- 'withRegistry', which itself is tied to the "parent registry" in which it was
-- created, this creates a hierarchy of registries.
```

Reply:
> Is that too thin of a scenario to accommodate without a concrete example to back it up? 

No, I don't think so, I thought about it also, but I guess I wasn't specific enough in the documentation. I've added the following part to the last section:

```
-- This includes 'withAsync': after all, 'withAsync' can't result in a child
-- thread that outlives its parent, so this is safe. In particular, if a parent
-- thread wants to link to a child thread /and handle any exceptions arising
-- from the link/, it should probably use `withAsync`:
--
-- > let handleLinkException :: ExceptionInLinkedThread -> m ()
-- >     handleLinkException = ..
-- > in handle handleLinkException $
-- >      withAsync codeInChild $ \child ->
-- >        ..
--
-- instead of
--
-- > handle handleLinkException $ do  -- PROBABLY NOT CORRECT!
-- >   child <- async codeInChild
-- >   ..
--
-- where the parent may exit the scope of the exception handler before the child
-- terminates. If the lifetime of the child cannot be limited to the lifetime of
-- the parent, the child should probably be linked to the registry instead and
-- the thread that spawned the registry should handle any exceptions.
```